spec/jobs/amend_backorder_job_spec.rb

```diff
@@ -0,0 +1,98 @@
+  let(:beans) { beans_item.variant }
+  let(:beans_item) { order.line_items[0] }
+  let(:chia_seed) { chia_item.variant }
+  let(:chia_item) { order.line_items[1] }
+  let(:user) { order.distributor.owner }
+  let(:catalog_json) { file_fixture("fdc-catalog.json").read }
+  let(:catalog_url) {
+    "https://env-0105831.jcloud-ver-jpe.ik-server.com/api/dfc/Enterprises/test-hodmedod/SuppliedProducts"
+  }
+  let(:product_link) {
+    "https://env-0105831.jcloud-ver-jpe.ik-server.com/api/dfc/Enterprises/test-hodmedod/SuppliedProducts/44519466467635"
+  }
+  let(:chia_seed_wholesale_link) {
+    "https://env-0105831.jcloud-ver-jpe.ik-server.com/api/dfc/Enterprises/test-hodmedod/SuppliedProducts/44519468433715"
+  }
+
+  before do
+    # This ensures that callbacks adjust stock correctly.
+    # See: https://github.com/openfoodfoundation/openfoodnetwork/pull/12938
+    order.reload
+
+    user.oidc_account = build(:testdfc_account)
+
+    beans.semantic_links << SemanticLink.new(
+      semantic_id: product_link
+    )
+    chia_seed.semantic_links << SemanticLink.new(
+      semantic_id: chia_seed_wholesale_link
+    )
+    order.order_cycle = create(
+      :simple_order_cycle,
+      distributors: [distributor],
+      variants: order.variants,
+    )
+    order.save!
+
+    beans.on_demand = true
+    beans_item.update!(quantity: 6)
+    beans.on_hand = -3
+
+    chia_item.update!(quantity: 5)
+    chia_seed.on_demand = false
+    chia_seed.on_hand = 7
+  end
+
+  describe "#amend_backorder" do
+    it "updates an order" do
+      stub_request(:get, catalog_url).to_return(body: catalog_json)
+
+      # Record the placed backorder:
+      backorder = nil
+      allow_any_instance_of(FdcBackorderer).to receive(:find_order) do |*_args|
+        backorder
+      end
+      allow_any_instance_of(FdcBackorderer).to receive(:find_open_order) do |*_args|
+        backorder
+      end
+      allow_any_instance_of(FdcBackorderer).to receive(:send_order) do |*args|
+        backorder = args[1]
+      end
+
+      BackorderJob.new.place_backorder(order)
+
+      # We ordered a case of 12 cans: -3 + 12 = 9
+      expect(beans.on_hand).to eq 9
+
+      # Stock controlled items don't change stock in backorder:
+      expect(chia_seed.on_hand).to eq 7
+
+      expect(backorder.lines[0].quantity).to eq 1 # beans
+      expect(backorder.lines[1].quantity).to eq 5 # chia
+
+      # Without any change, the backorder shouldn't get changed either:
+      subject.amend_backorder(order)
+
+      # Same as before:
+      expect(beans.on_hand).to eq 9
+      expect(chia_seed.on_hand).to eq 7
+      expect(backorder.lines[0].quantity).to eq 1 # beans
+      expect(backorder.lines[1].quantity).to eq 5 # chia
+
+      # We cancel the only order and that should reduce the order lines to 0.
+      expect { order.cancel! }
+        .to change { beans.reload.on_hand }.from(9).to(15)
+        .and change { chia_seed.reload.on_hand }.from(7).to(12)
```

Comment:
If the retail order and the backorder are both cancelled, why would the on_hand be changed?

Reply:
This block is cancelling the OFN order. So when you cancel the order the local OFN stock is returned. The backorder hasn't been touch yet. That's done in the next block.

When the backorder is updated, there are two different methods. If the product is on demand then the local stock may be reduced again because we are cancelling part of the backorder at the same time. But when it's stock controlled, we don't touch the OFN stock. The backorder is only affected by the line item quantities while the stock is synced with the remote catalog.